### PR TITLE
Return matched_queries in Percolator

### LINKED ISSUE #10163: Return `matched_queries` for named queries in percolator
I've got a bunch of multi-match percolator queries, which are named queries as well. It would be great to be able to get a response of matched queries including `matched_queries` when percolating. e.g.: 

```
  {
     "took": 0,
     "_shards": {
        "total": 5,
        "successful": 5,
        "failed": 0
     },
     "total": 2,
     "matches": [
        {
           "_index": "rethinkdb_ex",
           "_id": "user-1",
           "matched_queries": [
              "queryA"
           ]
        },
        {
           "_index": "rethinkdb_ex",
           "_id": "user-2",
           "matched_queries": [
              "queryA",
              "queryB"
           ]
        }
     ]
  }
```

Is this something that's planned?

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Return matched_queries for named queries in Percolator.

In a response, each hit together with
a `_percolator_document_slot` field will contain
`_percolator_document_slot_<slotNumber>_matched_queries` fields that will show which sub-queries matched each percolated document.

Closes #10163

### CODE REVIEW

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java [hunk 1]
```diff
@@ -96,7 +97,30 @@ public void process(HitContext hitContext) throws IOException {
 
                     IntStream slots = convertTopDocsToSlots(topDocs, pc.rootDocsBySlot);
                     // _percolator_document_slot fields are document fields and should be under "fields" section in a hit
-                    hitContext.hit().setDocumentField(fieldName, new DocumentField(fieldName, slots.boxed().collect(Collectors.toList())));
+                    List<Object> docSlots = slots.boxed().collect(Collectors.toList());
+                    hitContext.hit().setDocumentField(fieldName, new DocumentField(fieldName, docSlots));
+
+                    // Add info what sub-queries of percolator query matched this each percolated document
+                    if (fetchContext.getSearchExecutionContext().hasNamedQueries()) {
+                        List<LeafReaderContext> leafContexts = percolatorIndexSearcher.getLeafContexts();
```
Comment (benwtrent) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:105]: Why can't we use the `ctx` that is set via 
```
public void setNextReader(LeafReaderContext readerContext) {
                this.ctx = readerContext;
            }
```
?

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java [hunk 2]
```diff
@@ -96,7 +97,30 @@ public void process(HitContext hitContext) throws IOException {
...
+                            List<NamedMatches> namedMatchesList = NamedMatches.findNamedMatches(
+                                weight.matches(memoryReaderContext, topDocs.scoreDocs[i].doc)
+                            );
+                            if (namedMatchesList.isEmpty()) {
+                                continue;
+                            }
+                            List<Object> matchedQueries = new ArrayList<>(namedMatchesList.size());
+                            for (int j = 0; j < namedMatchesList.size(); j++) {
+                                matchedQueries.add(namedMatchesList.get(j).getName());
+                            }
```
Comment (benwtrent) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:119]:
```suggestion
                            List<Object> matchedQueries = new ArrayList<>(namedMatchesList.size());
                            for (NamedMatches match : namedMatchesList) {
                                matchedQueries.add(match.getName());
                            }
```
Reads cleaner to me. But minor.

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java [hunk 3]
```diff
@@ -96,7 +97,30 @@ public void process(HitContext hitContext) throws IOException {
...
+                        assert leafContexts.size() == 1 : "Expected single leaf, but got [" + leafContexts.size() + "]";
+                        LeafReaderContext memoryReaderContext = leafContexts.get(0);
+                        Weight weight = percolatorIndexSearcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1);
+                        for (int i = 0; i < topDocs.scoreDocs.length; i++) {
+                            List<NamedMatches> namedMatchesList = NamedMatches.findNamedMatches(
+                                weight.matches(memoryReaderContext, topDocs.scoreDocs[i].doc)
+                            );
+                            if (namedMatchesList.isEmpty()) {
+                                continue;
+                            }
```
Comment (benwtrent) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:115]: Oh wow! This makes me think this is how we should even do named query extraction in the `MatchedQueriesPhase`! I would think `matches` is way faster than us handrolling our own scorer iterators.

**Review by benwtrent — COMMENTED** [review]
Could you add tests to:
```
modules/percolator/src/yamlRestTest/resources/rest-api-spec/test/10_basic.yml
```

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java (hunk 1 above)
Comment (mayya-sharipova) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:105]: These are two different readerContext.

The one you suggested in `setNextReader` is for an index storing percolator queries.

But here, we need readerContext for an in-memory index storing percolating documents.

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java (hunk 2 above)
Comment (mayya-sharipova) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java]: Thanks, indeed it is cleaner. Addressed in b7a826d4feb118629ef977fa60abc9258b74876a

**Review by benwtrent — APPROVED** [review]
🪄

**Pull request merged**